runners/samza/src/main/java/org/apache/beam/runners/samza/renderer/PipelineJsonRenderer.java

```diff
@@ -0,0 +1,182 @@
+    ioInfo.ifPresent(s -> writeLine(" \"ioInfo\":\"%s\",", escapeString(s)));
+
+    writeLine("  \"ChildNodes\":[");
+    enterBlock();
+    return CompositeBehavior.ENTER_TRANSFORM;
+  }
+
+  @Override
+  public void leaveCompositeTransform(TransformHierarchy.Node node) {
+    exitBlock();
+    writeLine("]},");
+  }
+
+  @Override
+  public void visitPrimitiveTransform(TransformHierarchy.Node node) {
+    String fullName = node.getFullName();
+    writeLine("{ \"fullName\":\"%s\",", escapeString(fullName));
+    String enclosingNodeName = node.getEnclosingNode().getFullName();
+    writeLine("  \"enclosingNode\":\"%s\"},", assignNodeName(enclosingNodeName));
+
+    node.getOutputs().values().forEach(x -> valueToProducerNodeName.put(x, fullName));
+    node.getInputs()
+        .forEach(
+            (key, value) -> {
+              final String producerName = valueToProducerNodeName.get(value);
+              graphLinks.append(
+                  String.format("{\"from\":\"%s\"," + "\"to\":\"%s\"},", producerName, fullName));
+            });
+  }
+
+  @Override
+  public void visitValue(PValue value, TransformHierarchy.Node producer) {}
+
+  @Override
+  public void leavePipeline(Pipeline pipeline) {}
+
+  private void begin() {
+    writeLine("{ \n \"RootNode\": [");
+    graphLinks.append(",\"graphLinks\": [");
+    enterBlock();
+  }
+
+  private void end() {
+    exitBlock();
+    writeLine("]");
+    // delete the last comma
+    int lastIndex = graphLinks.length() - 1;
+    if (graphLinks.charAt(lastIndex) == ',') {
+      graphLinks.deleteCharAt(lastIndex);
+    }
+    graphLinks.append("]");
+    jsonBuilder.append(graphLinks);
+    jsonBuilder.append("}");
+  }
+
+  private void enterBlock() {
+    indent += 4;
+  }
+
+  private void exitBlock() {
+    indent -= 4;
+  }
+
+  private void writeLine(String format, Object... args) {
+    int secondLastCharIndex = jsonBuilder.length() - 2;
+    if (jsonBuilder.length() > 1
+        && jsonBuilder.charAt(secondLastCharIndex) == ','
+        && (format.startsWith("}") || format.startsWith("]"))) {
+      jsonBuilder.deleteCharAt(secondLastCharIndex);
+    }
+    if (indent != 0) {
+      jsonBuilder.append(String.format("%-" + indent + "s", ""));
+    }
+    jsonBuilder.append(String.format(format, args));
+    jsonBuilder.append("\n");
+  }
+
+  private static String escapeString(String x) {
+    return x.replace("\"", "\\\"");
+  }
+
+  private static String shortenTag(String tag) {
+    return tag.replaceFirst(".*:([a-zA-Z#0-9]+).*", "$1");
+  }
+
+  private String assignNodeName(String nodeName) {
+    return escapeString(nodeName.isEmpty() ? OUTERMOST_NODE : nodeName);
+  }
+
+  private Optional<String> getIOTopicInfo(TransformHierarchy.Node node) {
+    final SamzaIOInfo samzaIOInfo;
+    samzaIOInfo =
+        beamIORegistrarIterator.hasNext()
+            ? Iterators.getOnlyElement(beamIORegistrarIterator).getSamzaIO()
+            : null;
```

Comment:
I would expect this to be handled once instead of everytime visiting a node.

e.g. we can do

```
private static final SamzaIoInfo SAMZA_IO_INFO = loadSamzaIOInfo();

private static loadSamzaIOInfo() {
  final Iterator<SamzaIOInfo.SamzaIORegistrar> registrar/iterator = ServiceLoader.load(SamzaIOInfo.SamzaIORegistrar.class);
  return registrar/iterator.hasNext() ? Iterators.getOnlyElement(registrar/iterator) : null
}
```